internal/server/iigointernal/executive.go

```diff
@@ -47,75 +50,105 @@ func (e *executive) setAllocationRequest(resourceRequests map[shared.ClientID]sh
+				data[shared.TaxAmount] = d
+				communicateWithIslands(islandID, shared.TeamIDs[e.PresidentID], data)
+			}
 		}
 	}
+	return nil
 }
 
 // Send Tax map all the remaining islands
 // Called by orchestration at the end of the turn
-func (e *executive) getAllocationRequests(commonPool shared.Resources) (map[shared.ClientID]shared.Resources, bool) {
-	e.budget -= serviceCharge
+func (e *executive) getAllocationRequests(commonPool shared.Resources) shared.PresidentReturnContent {
 	return e.clientPresident.EvaluateAllocationRequests(e.ResourceRequests, commonPool)
 }
 
-func (e *executive) requestAllocationRequest() {
+func (e *executive) requestAllocationRequest(aliveIslands []shared.ClientID) error {
+	// Listening to islands requests incurs cost to the president. Set the cost to 0?
+	if !e.incurServiceCharge(e.gameConf.RequestAllocationRequestActionCost) {
+		return errors.Errorf("Insufficient Budget in common Pool: requestAllocationRequest")
+	}
 	allocRequests := make(map[shared.ClientID]shared.Resources)
-	for _, island := range getIslandAlive() {
-		allocRequests[shared.ClientID(int(island))] = iigoClients[shared.ClientID(int(island))].CommonPoolResourceRequest()
+	for _, islandID := range aliveIslands {
+		allocRequests[islandID] = iigoClients[islandID].CommonPoolResourceRequest()
 	}
 	AllocationAmountMapExport = allocRequests
 	e.setAllocationRequest(allocRequests)
+	return nil
 }
 
 // replyAllocationRequest broadcasts the allocation of resources decided by the president
 // to all islands alive
-func (e *executive) replyAllocationRequest(commonPool shared.Resources) bool {
-	e.budget -= serviceCharge
-	allocationMap, requestsEvaluated := e.getAllocationRequests(commonPool)
+func (e *executive) replyAllocationRequest(commonPool shared.Resources) (bool, error) {
+	if !CheckEnoughInCommonPool(e.gameConf.ReplyAllocationRequestsActionCost, e.gameState) {
+		return false, errors.Errorf("Insufficient Budget in common Pool: replyAllocationRequest")
+	}
+	returnContent := e.getAllocationRequests(commonPool)
 	allocationsMade := false
-	if requestsEvaluated {
+	if returnContent.ActionTaken {
+		if !e.incurServiceCharge(e.gameConf.ReplyAllocationRequestsActionCost) {
+			return false, errors.Errorf("Insufficient Budget in common Pool: replyAllocationRequest")
+		}
 		allocationsMade = true
 		for _, island := range getIslandAlive() {
-			d := shared.CommunicationContent{T: shared.CommunicationInt, IntegerData: int(allocationMap[shared.ClientID(int(island))])}
+			d := shared.CommunicationContent{T: shared.CommunicationInt, IntegerData: int(returnContent.ResourceMap[shared.ClientID(int(island))])}
 			data := make(map[shared.CommunicationFieldName]shared.CommunicationContent)
 			data[shared.AllocationAmount] = d
-			communicateWithIslands(shared.TeamIDs[int(island)], shared.TeamIDs[e.ID], data)
+			communicateWithIslands(shared.TeamIDs[int(island)], shared.TeamIDs[int(e.PresidentID)], data)
```

Comment:
Could we use shared.ClientID(island) instead of shared.TeamIDs[int(island)], the latter could cause an index error and panic rather than be caught by the check in communicateWithIslands